File: internal/sdk/typescript/src/datasets.ts
import { create, type MessageInitShape } from "@bufbuild/protobuf";

import {
	type ResourceRef,
	ResourceRefSchema,
} from "../../../../protocols/generated/typescript/common/v1/resource_reference_pb.js";
import {
	CreateDatasetCommandSchema,
	PublishDatasetReleaseCommandSchema,
	RevokeDatasetReleaseCommandSchema,
	UpdateDatasetCommandSchema,
} from "../../../../protocols/generated/typescript/dataset/v1/dataset_commands_pb.js";
import type { Dataset } from "../../../../protocols/generated/typescript/dataset/v1/dataset_pb.js";
import type { DatasetRelease } from "../../../../protocols/generated/typescript/dataset/v1/dataset_release_pb.js";
import {
	CreateDatasetRequestSchema,
	GetDatasetReleaseRequestSchema,
	GetDatasetRequestSchema,
	ListDatasetReleasesRequestSchema,
	type ListDatasetReleasesResponse,
	ListDatasetsRequestSchema,
	type ListDatasetsResponse,
	PublishDatasetReleaseRequestSchema,
	RevokeDatasetReleaseRequestSchema,
	UpdateDatasetRequestSchema,
} from "../../../../protocols/generated/typescript/internal/dataset/v1/dataset_service_pb.js";
import type { Operation } from "../../../../protocols/generated/typescript/job/v1/operation_pb.js";
import type { ClientCore } from "./core.js";
import { MindcladeError } from "./error.js";
import { commandContext, prepareCall, type SdkCallOptions, type SubmitOptions } from "./request.js";
import { invokeUnary } from "./retry.js";
import { registeredMethodSafety } from "./safety.js";

const CREATE = "/mindclade.internal.dataset.v1.DatasetService/CreateDataset";
const GET = "/mindclade.internal.dataset.v1.DatasetService/GetDataset";
const LIST = "/mindclade.internal.dataset.v1.DatasetService/ListDatasets";
const UPDATE = "/mindclade.internal.dataset.v1.DatasetService/UpdateDataset";
const PUBLISH = "/mindclade.internal.dataset.v1.DatasetService/PublishDatasetRelease";
const REVOKE = "/mindclade.internal.dataset.v1.DatasetService/RevokeDatasetRelease";
const GET_RELEASE = "/mindclade.internal.dataset.v1.DatasetService/GetDatasetRelease";
const LIST_RELEASES = "/mindclade.internal.dataset.v1.DatasetService/ListDatasetReleases";

/** Private Dataset lifecycle facade. Every request and response is generated. */
export class Datasets {
	readonly #core: ClientCore;

	constructor(core: ClientCore) {
		this.#core = core;
	}

	async create(
		command: MessageInitShape<typeof CreateDatasetCommandSchema>,
		options: SubmitOptions,
	): Promise<Operation> {
		const parent = projectName(this.#core);
		if (command.project?.name !== undefined && command.project.name !== parent) {
			throw MindcladeError.invalidArgument("dataset project does not match client scope");
		}
		const prepared = prepareCall(this.#core.config, this.#core.runtime, options);
		const generated = create(CreateDatasetCommandSchema, command);
		delete generated.context;
		generated.project ??= projectRef(this.#core);
		// JavaScript map iteration is not a cross-language canonical encoding.
		// The authoritative service validates this identity context and persists
		// the deterministic request digest it computes from the generated command.
		generated.context = commandContext(this.#core.config, prepared, options);
		const response = await invokeUnary(
			this.#core,
			prepared,
			registeredMethodSafety(CREATE),
			options.idempotencyKey,
			(call) =>
				this.#core.raw.datasets.createDataset(
					create(CreateDatasetRequestSchema, { command: generated }),
					call,
				),
		);
		return requiredOperation(response.operation, "CreateDataset");
	}

	async get(
		request: MessageInitShape<typeof GetDatasetRequestSchema>,
		options: SdkCallOptions = {},
	): Promise<Dataset> {
		const generated = create(GetDatasetRequestSchema, request);
		generated.name = datasetName(this.#core, generated.name);
		const prepared = prepareCall(this.#core.config, this.#core.runtime, options);
		const response = await invokeUnary(
			this.#core,
			prepared,
			registeredMethodSafety(GET),
			undefined,
			(call) => this.#core.raw.datasets.getDataset(generated, call),
		);
		if (response.dataset === undefined)
			throw MindcladeError.protocol("GetDataset response omitted its dataset");
		return response.dataset;
	}

	async list(
		request: MessageInitShape<typeof ListDatasetsRequestSchema> = {},
		options: SdkCallOptions = {},
	): Promise<ListDatasetsResponse> {
		const generated = create(ListDatasetsRequestSchema, request);
		const parent = projectName(this.#core);
		if (generated.parent !== "" && generated.parent !== parent)
			throw MindcladeError.invalidArgument("dataset list parent does not match client scope");
		generated.parent = parent;
		validatePage(generated.page?.pageSize);
		const prepared = prepareCall(this.#core.config, this.#core.runtime, options);
		return await invokeUnary(
			this.#core,
			prepared,
			registeredMethodSafety(LIST),
			undefined,
			(call) => this.#core.raw.datasets.listDatasets(generated, call),
		);
	}

	async update(
		command: MessageInitShape<typeof UpdateDatasetCommandSchema>,
		options: SubmitOptions,
	): Promise<Operation> {
		if (command.dataset?.name === undefined)
			throw MindcladeError.invalidArgument("update requires a generated dataset");
		datasetName(this.#core, command.dataset.name);
		const prepared = prepareCall(this.#core.config, this.#core.runtime, options);
		const generated = create(UpdateDatasetCommandSchema, command);
		delete generated.context;
		generated.context = commandContext(this.#core.config, prepared, options);
		const response = await invokeUnary(
			this.#core,
			prepared,
			registeredMethodSafety(UPDATE),
			options.idempotencyKey,
			(call) =>
				this.#core.raw.datasets.updateDataset(
					create(UpdateDatasetRequestSchema, { command: generated }),
					call,
				),
		);
		return requiredOperation(response.operation, "UpdateDataset");
	}

	async publishRelease(
		command: MessageInitShape<typeof PublishDatasetReleaseCommandSchema>,
		options: SubmitOptions,
	): Promise<Operation> {
		if (command.dataset?.name === undefined)
			throw MindcladeError.invalidArgument("publication requires a generated dataset reference");
		datasetName(this.#core, command.dataset.name);
		const prepared = prepareCall(this.#core.config, this.#core.runtime, options);
		const generated = create(PublishDatasetReleaseCommandSchema, command);
		delete generated.context;
		if (generated.dataset === undefined)
			throw MindcladeError.invalidArgument("publication requires a generated dataset reference");
		normalizeReference(this.#core, generated.dataset, "dataset", false);
		generated.context = commandContext(this.#core.config, prepared, options);
		const response = await invokeUnary(
			this.#core,
			prepared,
			registeredMethodSafety(PUBLISH),
			options.idempotencyKey,
			(call) =>
				this.#core.raw.datasets.publishDatasetRelease(
					create(PublishDatasetReleaseRequestSchema, { command: generated }),
					call,
				),
		);
		return requiredOperation(response.operation, "PublishDatasetRelease");
	}

	async revokeRelease(
		command: MessageInitShape<typeof RevokeDatasetReleaseCommandSchema>,
		options: SubmitOptions,
	): Promise<Operation> {
		if (command.datasetRelease?.name === undefined)
			throw MindcladeError.invalidArgument(
				"revocation requires a generated dataset release reference",
			);
		releaseName(this.#core, command.datasetRelease.name);
		const prepared = prepareCall(this.#core.config, this.#core.runtime, options);
		const generated = create(RevokeDatasetReleaseCommandSchema, command);
		delete generated.context;
		if (generated.datasetRelease === undefined)
			throw MindcladeError.invalidArgument(
				"revocation requires a generated dataset release reference",
			);
		normalizeReference(this.#core, generated.datasetRelease, "dataset_release", true);
		generated.context = commandContext(this.#core.config, prepared, options);
		const response = await invokeUnary(
			this.#core,
			prepared,
			registeredMethodSafety(REVOKE),
			options.idempotencyKey,
			(call) =>
				this.#core.raw.datasets.revokeDatasetRelease(
					create(RevokeDatasetReleaseRequestSchema, { command: generated }),
					call,
				),
		);
		return requiredOperation(response.operation, "RevokeDatasetRelease");
	}

	async getRelease(
		request: MessageInitShape<typeof GetDatasetReleaseRequestSchema>,
		options: SdkCallOptions = {},
	): Promise<DatasetRelease> {
		const generated = create(GetDatasetReleaseRequestSchema, request);
		generated.name = releaseName(this.#core, generated.name);
		const prepared = prepareCall(this.#core.config, this.#core.runtime, options);
		const response = await invokeUnary(
			this.#core,
			prepared,
			registeredMethodSafety(GET_RELEASE),
			undefined,
			(call) => this.#core.raw.datasets.getDatasetRelease(generated, call),
		);
		if (response.datasetRelease === undefined)
			throw MindcladeError.protocol("GetDatasetRelease response omitted its release");
		return response.datasetRelease;
	}

	async listReleases(
		request: MessageInitShape<typeof ListDatasetReleasesRequestSchema>,
		options: SdkCallOptions = {},
	): Promise<ListDatasetReleasesResponse> {
		const generated = create(ListDatasetReleasesRequestSchema, request);
		generated.parent = datasetName(this.#core, generated.parent);
		validatePage(generated.page?.pageSize);
		const prepared = prepareCall(this.#core.config, this.#core.runtime, options);
		return await invokeUnary(
			this.#core,
			prepared,
			registeredMethodSafety(LIST_RELEASES),
			undefined,
			(call) => this.#core.raw.datasets.listDatasetReleases(generated, call),
		);
	}
}

const projectName = (core: ClientCore): string => {
	const tenant = core.config.identity.tenantId.startsWith("tenants/")
		? core.config.identity.tenantId
		: `tenants/${core.config.identity.tenantId}`;
	const project = core.config.identity.projectId;
	if (project.startsWith("tenants/")) return project;
	return project.startsWith("projects/") ? `${tenant}/${project}` : `${tenant}/projects/${project}`;
};
const projectRef = (core: ClientCore) =>
	create(ResourceRefSchema, {
		resourceType: "project",
		resourceId: core.config.identity.projectId,
		tenantId: core.config.identity.tenantId,
		projectId: core.config.identity.projectId,
		name: projectName(core),
	});

const datasetName = (core: ClientCore, name: string): string => scopedName(core, name, false);
const releaseName = (core: ClientCore, name: string): string => scopedName(core, name, true);
const scopedName = (core: ClientCore, name: string, release: boolean): string => {
	const prefix = `${projectName(core)}/datasets/`;
	if (!name.startsWith(prefix))
		throw MindcladeError.invalidArgument("resource is outside the configured project");
	const suffix = name.slice(prefix.length);
	const valid = release ? /^[^/]+\/releases\/[^/]+$/.test(suffix) : /^[^/]+$/.test(suffix);
	if (!valid) throw MindcladeError.invalidArgument("resource name is invalid");
	return name;
};
const normalizeReference = (
	core: ClientCore,
	reference: ResourceRef,
	resourceType: string,
	release: boolean,
): void => {
	const name = scopedName(core, reference.name, release);
	const resourceId = name.slice(name.lastIndexOf("/") + 1);
	if (
		(reference.resourceType !== "" && reference.resourceType !== resourceType) ||
		(reference.resourceId !== "" && reference.resourceId !== resourceId) ||
		(reference.tenantId !== "" && reference.tenantId !== core.config.identity.tenantId) ||
		(reference.projectId !== "" && reference.projectId !== core.config.identity.projectId)
	) {
		throw MindcladeError.invalidArgument(
			"resource reference does not match the configured project",
		);
	}
	reference.resourceType = resourceType;
	reference.resourceId = resourceId;
	reference.tenantId = core.config.identity.tenantId;
	reference.projectId = core.config.identity.projectId;
};
const validatePage = (size: number | undefined): void => {
	if (size !== undefined && (!Number.isInteger(size) || size < 0 || size > 1000))
		throw MindcladeError.invalidArgument("page size must be an integer between zero and 1000");
};
const requiredOperation = (operation: Operation | undefined, method: string): Operation => {
	if (operation === undefined || operation.operationId.trim() === "")
		throw MindcladeError.protocol(`${method} response omitted its durable operation`);
	return operation;
};

File: internal/sdk/rust/src/datasets.rs
use std::sync::Arc;

use mindclade_protocols::{
    common::v1::ResourceRef,
    dataset::v1::{
        CreateDatasetCommand, Dataset, DatasetRelease, PublishDatasetReleaseCommand,
        RevokeDatasetReleaseCommand, UpdateDatasetCommand,
    },
    internal::dataset::v1::{
        CreateDatasetRequest, GetDatasetReleaseRequest, GetDatasetRequest,
        ListDatasetReleasesRequest, ListDatasetsRequest, PublishDatasetReleaseRequest,
        RevokeDatasetReleaseRequest, UpdateDatasetRequest,
    },
    job::v1::Operation,
};

use crate::{
    CallOptions, ClientCore, Error, Page, Pages, SubmitOptions,
    request::{initial_page_token, page_request},
    retry::registered_method_safety,
};

const CREATE: &str = "/mindclade.internal.dataset.v1.DatasetService/CreateDataset";
const GET: &str = "/mindclade.internal.dataset.v1.DatasetService/GetDataset";
const LIST: &str = "/mindclade.internal.dataset.v1.DatasetService/ListDatasets";
const UPDATE: &str = "/mindclade.internal.dataset.v1.DatasetService/UpdateDataset";
const PUBLISH: &str = "/mindclade.internal.dataset.v1.DatasetService/PublishDatasetRelease";
const REVOKE: &str = "/mindclade.internal.dataset.v1.DatasetService/RevokeDatasetRelease";
const GET_RELEASE: &str = "/mindclade.internal.dataset.v1.DatasetService/GetDatasetRelease";
const LIST_RELEASES: &str = "/mindclade.internal.dataset.v1.DatasetService/ListDatasetReleases";

/// Private Dataset lifecycle API backed exclusively by generated contracts.
#[derive(Clone)]
pub struct Datasets {
    core: Arc<ClientCore>,
}

impl Datasets {
    pub(crate) fn new(core: Arc<ClientCore>) -> Self {
        Self { core }
    }

    /// Creates a dataset and returns its durable generated operation.
    ///
    /// # Errors
    ///
    /// Returns an error for invalid scope, command metadata, credentials, transport failure, or a malformed response.
    pub async fn create(
        &self,
        mut command: CreateDatasetCommand,
        options: SubmitOptions,
    ) -> Result<Operation, Error> {
        let expected = project_name(&self.core);
        match &command.project {
            Some(project) if project.name != expected => {
                return Err(Error::invalid_argument(
                    "dataset project does not match client scope",
                ));
            }
            None => command.project = Some(project_ref(&self.core)),
            _ => {}
        }
        command.context = None;
        let prepared = options.call.prepare(&self.core.config);
        command.context = Some(prepared.command_context(&self.core.config, &options)?);
        let key = options.idempotency_key.clone();
        let response = self
            .core
            .unary(
                CreateDatasetRequest {
                    command: Some(command),
                },
                &prepared,
                registered_method_safety(CREATE),
                Some(&key),
                |transport, request| {
                    Box::pin(async move { transport.create_dataset(request).await })
                },
            )
            .await?
            .into_inner();
        require_operation(response.operation, "CreateDataset")
    }

    /// Gets one generated dataset resource in the configured project.
    ///
    /// # Errors
    ///
    /// Returns an error for an invalid name, credentials, transport failure, or a malformed response.
    pub async fn get(
        &self,
        name: impl Into<String>,
        if_none_match: impl Into<String>,
        options: CallOptions,
    ) -> Result<Dataset, Error> {
        let name = dataset_name(&self.core, &name.into())?;
        let prepared = options.prepare(&self.core.config);
        let response = self
            .core
            .unary(
                GetDatasetRequest {
                    name,
                    if_none_match: if_none_match.into(),
                },
                &prepared,
                registered_method_safety(GET),
                None,
                |transport, request| Box::pin(async move { transport.get_dataset(request).await }),
            )
            .await?
            .into_inner();
        response
            .dataset
            .ok_or_else(|| Error::protocol("GetDataset response omitted its dataset"))
    }

    /// Lists one bounded page of generated dataset resources.
    ///
    /// # Errors
    ///
    /// Returns an error for invalid scope or pagination, credentials, or transport failure.
    pub fn list(
        &self,
        mut request: ListDatasetsRequest,
        options: CallOptions,
    ) -> Result<Pages<Dataset>, Error> {
        let parent = project_name(&self.core);
        if !request.parent.is_empty() && request.parent != parent {
            return Err(Error::invalid_argument(
                "dataset list parent does not match client scope",
            ));
        }
        request.parent = parent;
        validate_page(request.page.as_ref())?;
        let core = Arc::clone(&self.core);
        let token = initial_page_token(request.page.as_ref());
        Ok(Pages::new(
            move |page_token| {
                let core = Arc::clone(&core);
                let options = options.clone();
                let mut request = request.clone();
                async move {
                    request.page = Some(page_request(request.page.as_ref(), page_token));
                    let prepared = options.prepare(&core.config);
                    let response = core
                        .unary(
                            request,
                            &prepared,
                            registered_method_safety(LIST),
                            None,
                            |transport, request| {
                                Box::pin(async move { transport.list_datasets(request).await })
                            },
                        )
                        .await?;
                    let request_id = response.request_id().map(str::to_owned);
                    let response = response.into_inner();
                    Ok(Page::new(
                        response.datasets,
                        response.page,
                        response.read_time,
                        request_id,
                    ))
                }
            },
            token,
        ))
    }

    /// Applies a generated optimistic dataset update command.
    ///
    /// # Errors
    ///
    /// Returns an error for invalid scope or command metadata, credentials, transport failure, or a malformed response.
    pub async fn update(
        &self,
        mut command: UpdateDatasetCommand,
        options: SubmitOptions,
    ) -> Result<Operation, Error> {
        dataset_name(
            &self.core,
            &command
                .dataset
                .as_ref()
                .ok_or_else(|| Error::invalid_argument("update requires a dataset"))?
                .name,
        )?;
        command.context = None;
        let prepared = options.call.prepare(&self.core.config);
        command.context = Some(prepared.command_context(&self.core.config, &options)?);
        let key = options.idempotency_key.clone();
        let response = self
            .core
            .unary(
                UpdateDatasetRequest {
                    command: Some(command),
                },
                &prepared,
                registered_method_safety(UPDATE),
                Some(&key),
                |transport, request| {
                    Box::pin(async move { transport.update_dataset(request).await })
                },
            )
            .await?
            .into_inner();
        require_operation(response.operation, "UpdateDataset")
    }

    /// Publishes an immutable dataset release through a durable operation.
    ///
    /// # Errors
    ///
    /// Returns an error for an invalid reference or command, credentials, transport failure, or a malformed response.
    pub async fn publish_release(
        &self,
        mut command: PublishDatasetReleaseCommand,
        options: SubmitOptions,
    ) -> Result<Operation, Error> {
        normalize_reference(
            &self.core,
            command.dataset.as_mut().ok_or_else(|| {
                Error::invalid_argument("publication requires a dataset reference")
            })?,
            "dataset",
            false,
        )?;
        command.context = None;
        let prepared = options.call.prepare(&self.core.config);
        command.context = Some(prepared.command_context(&self.core.config, &options)?);
        let key = options.idempotency_key.clone();
        let response = self
            .core
            .unary(
                PublishDatasetReleaseRequest {
                    command: Some(command),
                },
                &prepared,
                registered_method_safety(PUBLISH),
                Some(&key),
                |transport, request| {
                    Box::pin(async move { transport.publish_dataset_release(request).await })
                },
            )
            .await?
            .into_inner();
        require_operation(response.operation, "PublishDatasetRelease")
    }

    /// Records a non-destructive dataset release revocation.
    ///
    /// # Errors
    ///
    /// Returns an error for an invalid reference or command, credentials, transport failure, or a malformed response.
    pub async fn revoke_release(
        &self,
        mut command: RevokeDatasetReleaseCommand,
        options: SubmitOptions,
    ) -> Result<Operation, Error> {
        normalize_reference(
            &self.core,
            command.dataset_release.as_mut().ok_or_else(|| {
                Error::invalid_argument("revocation requires a dataset release reference")
            })?,
            "dataset_release",
            true,
        )?;
        command.context = None;
        let prepared = options.call.prepare(&self.core.config);
        command.context = Some(prepared.command_context(&self.core.config, &options)?);
        let key = options.idempotency_key.clone();
        let response = self
            .core
            .unary(
                RevokeDatasetReleaseRequest {
                    command: Some(command),
                },
                &prepared,
                registered_method_safety(REVOKE),
                Some(&key),
                |transport, request| {
                    Box::pin(async move { transport.revoke_dataset_release(request).await })
                },
            )
            .await?
            .into_inner();
        require_operation(response.operation, "RevokeDatasetRelease")
    }

    /// Gets one generated immutable dataset release.
    ///
    /// # Errors
    ///
    /// Returns an error for an invalid name, credentials, transport failure, or a malformed response.
    pub async fn get_release(
        &self,
        name: impl Into<String>,
        options: CallOptions,
    ) -> Result<DatasetRelease, Error> {
        let name = dataset_release_name(&self.core, &name.into())?;
        let prepared = options.prepare(&self.core.config);
        let response = self
            .core
            .unary(
                GetDatasetReleaseRequest { name },
                &prepared,
                registered_method_safety(GET_RELEASE),
                None,
                |transport, request| {
                    Box::pin(async move { transport.get_dataset_release(request).await })
                },
            )
            .await?
            .into_inner();
        response
            .dataset_release
            .ok_or_else(|| Error::protocol("GetDatasetRelease response omitted its release"))
    }

    /// Lists one bounded page of releases under a generated dataset name.
    ///
    /// # Errors
    ///
    /// Returns an error for invalid scope or pagination, credentials, or transport failure.
    pub fn list_releases(
        &self,
        request: ListDatasetReleasesRequest,
        options: CallOptions,
    ) -> Result<Pages<DatasetRelease>, Error> {
        dataset_name(&self.core, &request.parent)?;
        validate_page(request.page.as_ref())?;
        let core = Arc::clone(&self.core);
        let token = initial_page_token(request.page.as_ref());
        Ok(Pages::new(
            move |page_token| {
                let core = Arc::clone(&core);
                let options = options.clone();
                let mut request = request.clone();
                async move {
                    request.page = Some(page_request(request.page.as_ref(), page_token));
                    let prepared = options.prepare(&core.config);
                    let response = core
                        .unary(
                            request,
                            &prepared,
                            registered_method_safety(LIST_RELEASES),
                            None,
                            |transport, request| {
                                Box::pin(
                                    async move { transport.list_dataset_releases(request).await },
                                )
                            },
                        )
                        .await?;
                    let request_id = response.request_id().map(str::to_owned);
                    let response = response.into_inner();
                    Ok(Page::new(
                        response.dataset_releases,
                        response.page,
                        response.read_time,
                        request_id,
                    ))
                }
            },
            token,
        ))
    }
}

fn project_name(core: &ClientCore) -> String {
    project_parent(&core.config)
}

fn project_parent(config: &crate::Config) -> String {
    let tenant = if config.identity.tenant_id().starts_with("tenants/") {
        config.identity.tenant_id().to_owned()
    } else {
        format!("tenants/{}", config.identity.tenant_id())
    };
    let project = config.identity.project_id();
    if project.starts_with("tenants/") {
        project.to_owned()
    } else if project.starts_with("projects/") {
        format!("{tenant}/{project}")
    } else {
        format!("{tenant}/projects/{project}")
    }
}

fn project_ref(core: &ClientCore) -> ResourceRef {
    ResourceRef {
        resource_type: "project".to_owned(),
        resource_id: core.config.identity.project_id().to_owned(),
        tenant_id: core.config.identity.tenant_id().to_owned(),
        project_id: core.config.identity.project_id().to_owned(),
        name: project_name(core),
        ..ResourceRef::default()
    }
}

fn dataset_name(core: &ClientCore, name: &str) -> Result<String, Error> {
    scoped_name(core, name, "datasets", false)
}

fn dataset_release_name(core: &ClientCore, name: &str) -> Result<String, Error> {
    scoped_name(core, name, "datasets", true)
}

fn scoped_name(
    core: &ClientCore,
    name: &str,
    collection: &str,
    release: bool,
) -> Result<String, Error> {
    let prefix = format!("{}/{collection}/", project_name(core));
    let suffix = name
        .strip_prefix(&prefix)
        .ok_or_else(|| Error::invalid_argument("resource is outside the configured project"))?;
    let valid = if release {
        let mut parts = suffix.split("/releases/");
        matches!((parts.next(), parts.next(), parts.next()), (Some(left), Some(right), None) if !left.is_empty() && !right.is_empty() && !left.contains('/') && !right.contains('/'))
    } else {
        !suffix.is_empty() && !suffix.contains('/')
    };
    if !valid {
        return Err(Error::invalid_argument("resource name is invalid"));
    }
    Ok(name.to_owned())
}

fn normalize_reference(
    core: &ClientCore,
    reference: &mut ResourceRef,
    resource_type: &str,
    release: bool,
) -> Result<(), Error> {
    let name = if release {
        dataset_release_name(core, &reference.name)?
    } else {
        dataset_name(core, &reference.name)?
    };
    let resource_id = name
        .rsplit('/')
        .next()
        .ok_or_else(|| Error::invalid_argument("resource reference name is invalid"))?;
    if (!reference.resource_type.is_empty() && reference.resource_type != resource_type)
        || (!reference.resource_id.is_empty() && reference.resource_id != resource_id)
        || (!reference.tenant_id.is_empty()
            && reference.tenant_id != core.config.identity.tenant_id())
        || (!reference.project_id.is_empty()
            && reference.project_id != core.config.identity.project_id())
    {
        return Err(Error::invalid_argument(
            "resource reference does not match the configured project",
        ));
    }
    reference.resource_type.clear();
    reference.resource_type.push_str(resource_type);
    reference.resource_id.clear();
    reference.resource_id.push_str(resource_id);
    reference.tenant_id.clear();
    reference
        .tenant_id
        .push_str(core.config.identity.tenant_id());
    reference.project_id.clear();
    reference
        .project_id
        .push_str(core.config.identity.project_id());
    Ok(())
}

fn validate_page(page: Option<&mindclade_protocols::common::v1::PageRequest>) -> Result<(), Error> {
    if page.is_some_and(|value| value.page_size > 1000) {
        return Err(Error::invalid_argument("page size cannot exceed 1000"));
    }
    Ok(())
}

fn require_operation(operation: Option<Operation>, method: &str) -> Result<Operation, Error> {
    let value = operation
        .ok_or_else(|| Error::protocol(format!("{method} response omitted its operation")))?;
    if value.operation_id.trim().is_empty() {
        return Err(Error::protocol(format!(
            "{method} response operation has no identity"
        )));
    }
    Ok(value)
}
